core/lib/instrument_api.h

```diff
@@ -1801,11 +1801,63 @@ DR_API
 bool
 dr_set_client_version_string(const char *version);
 
+DR_API
+/**
+ * Returns the error code of the last failed API routine. Users should check whether or
+ * not the API routine that they just called has failed prior to this function.
+ *
+ * \warning Not all API routines currently support the registering of an error code upon
+ * their failure. Therefore, check the routine's documentation to see whether it supports
+ * setting error codes.
+ */
+dr_error_code_t
+dr_get_error_code(void);
+
 DR_API
 /** Returns the image name (without path) of the current application. */
 const char *
 dr_get_application_name(void);
 
+DR_API
+/**
+ * Provides information about the app's arguments by setting \p args_buf up to
+ * the size \p buf_size. Returns the number of args set or -1 on error.
+ *
+ * Use dr_app_arg_as_cstring() to get the argument as a string.
+ *
+ * \note Currently, this function is only available on Unix with
+ * early injection.
+ *
+ * \note An error code may be obtained via dr_get_error_code() when this routine fails.
+ */
+int
+dr_get_app_args(OUT dr_app_arg_t *args_buf, int buf_size);
+
+DR_API
+/* Returns the number of command-line arguments passed to the application.
+ *
+ * \note Currently, this function is only available on Unix with
+ * early injection.
+ */
+int
+dr_num_app_args(void);
+
+DR_API
+/**
+ * Returns the passed argument \p app_arg as a string. \p buf is used only if needed, and
+ * therefore the caller should not assume that the string is in the \p buf. In other
+ * words, always use the returned value to refer to the string. Returns NULL on error
+ * such as when \p buf is needed as storage and the size of the buffer \p buf_size
+ * is not sufficient.
```

Comment:
I would mention how to get an upper bound on the size needed

Reply:
Done.